### ISSUE TITLE
Save only the best checkpoint

### ISSUE THREAD

**Issue by bernardohenz** [mozilla/DeepSpeech#1475]
Instead of using EarlyStop to avoid overfitting, I would like to save the model at the time it had the best (lowest) validation loss. In other words, I would like to keep track only for the best checkpoint (based on val/dev set).

I tried to use a wrapper for tf.train.Saver (this one: https://github.com/vonclites/checkmate), but couldn't make it works with DeepSpeech. Is there a easy way to do tha (maybe using the MonitoredTrainingSession as you are using)?

**Comment by reuben** [comment]
I looked into this briefly but couldn't find a clean way to implement it with MonitoredTrainingSession (which is IMO a terrible API). I ended up just writing a hack that works, but isn't really code we can land. I'm attaching the patch.

[save_best_val.patch.txt](https://github.com/mozilla/DeepSpeech/files/2225514/save_best_val.patch.txt)

**Comment by bernardohenz** [comment]
Thanks @reuben , I had to change the name of the MonitoredTrainingSession to `train_session` for this to work. Now it is working perfectly fine.

**Comment by lissyx** [comment]
@bernardohenz What's the status here, is the issue fixed, do you have a workaround ? Should we close this ?

**Comment by bernardohenz** [comment]
@lissyx yes, the patch from @reuben worked just fine.

**Comment by reuben** [comment]
We should have a proper solution for this in-tree. This would be too much work with the current training setup, but would probably be very simple if we used TF Eager, for example. Reopening so we don't forget.

**Comment by reuben** [comment]
Updated version of the patch is here: https://gist.github.com/reuben/dcc2deaf85568591e34ce363bc3bac2a

**Comment by tilmankamp** [comment]
#1988 added this feature. It checkpoints the best validating epoch and supports different ways to resume - see `load` command-line parameter.
